### DISCUSSION TITLE
Starmap Editor

### DISCUSSION THREAD

**Discussion by alexanderson1993** [Thorium-Sim/thorium-nova#159] [Ideas]
Okay, I think the first feature that absolutely needs to be added should be the star map editor. I really liked what I did for Thorium Nova Prime in 2020, so I'll replicate most of that behavior and copy a good portion of the code from that. Check the blog entries to get an idea for how that worked.

Here's the rough outline for how it will work. 

- The star map is built into plugins. When you run a flight with multiple plugins activated, the star map entities are all created and combined together.
- Ultimately, the star map consists of solar systems (or interstellar points of interest - areas that can be visited in interstellar space that have no star), stars, planets, moons, and permanent ships like starbases. 
- In the future, but not now, there will also be comets, asteroid belts, nebulas, and other stellar phenomena
- There will be a stage for interstellar objects (just solar systems) and a stage for all the objects within a solar system. The user experience involves selecting a solar system, entering it to see all of its contents, and then clicking a "Back" button to return to the interstellar stage.
- Both stages (interstellar and solar) will have a 2D top-down view (the default view) and a 3D view. The goal is to encourage a flat XZ plane of solar systems and planets, with the ability to add some Y (height) variation. 
- In the interstellar 2D view, entities are added to the Y=0 plane and can be configured to have a different Y position manually. Dragging entities changes their position on the XZ plane. 
- In the interstellar 3D view, entities are added some distance in front of the cameras view based on the current zoom level. Dragging entities changes their position along a plane perpendicular to the direction the camera is facing.
-  Stars, planets, and moons are positioned entirely based on their orbital properties. For stars, they are placed in the center of the system with the appropriate distance from their binary partners, if any. For planets, they are placed based on their orbit radius, eccentricity (how squashed the ellipse is), the orbit period (how far along the orbit the planet is) and the inclination (vertical tilt). Moons are the same as planets, except their origin is the center of their parent planet.
- Both views will have really nice zooming, panning, and rotating in the case of the 3D view.
- Entities can be selected. A panel (either an overlay palette or a slide out panel, I can't decide which is better) will allow the properties of the selected entity to be edited.
- Multiple entities can be selected and moved simultaneously. Cut/Copy/Paste keyboard shortcuts should be available.

I think that's enough to start. Anybody have any thoughts?

**Comment by astrolamb-gaming** [comment]
I think this is pretty well thought out so far. We may need a few more orbital properties for planetary bodies (argument of periapsis, actually here's the [list ](https://en.wikipedia.org/wiki/Orbital_elements))

**Reply by alexanderson1993** [reply]
Nice list. Thanks for sharing.

I wanted to keep it as simple as possible, so I used default values of 0 for the argument of periapsis and Longitude of the ascending node. Perhaps we could still track and store those values, but have them be advanced features that aren't available by default.
